### PR TITLE
Hessian-vector product of the residual with AD

### PR DESCRIPTION
**Pull request by kliegeois**
This PR adds the capability to use Automatic Differentiation (AD) to compute the Hessian-vector products of the residual by nesting AD types.

This PR adds four contributions of the Hessian-vector products to the previously merged ones:
- ,
- ,
- ,
- , 

where  is the solution,  is a first distributed parameter,  is a potentially different second distributed parameter,  is the residual,  is the Lagrange multiplier vector,  is a direction vector with the same dimension as the vector ,  is a direction vector with the same dimension as the vector , and  is a direction vector with the same dimension as the vector .

The implementation has been tested comparing results computed with AD and with Finite Differences (FD) of the gradients on a heat problem and FO_GIS examples.
However, this PR does not include tests yet. They will be included in a second PR.

### CODE REVIEW

File: src/LandIce/evaluators/LandIce_Gather2DField_Def.hpp [hunk 1]
```diff
@@ -357,8 +363,14 @@ evaluateFields(typename Traits::EvalData workset)
   bool g_xx_is_active = !workset.hessianWorkset.hess_vec_prod_g_xx.is_null();
   bool g_xp_is_active = !workset.hessianWorkset.hess_vec_prod_g_xp.is_null();
   bool g_px_is_active = !workset.hessianWorkset.hess_vec_prod_g_px.is_null();
+  bool f_xx_is_active = !workset.hessianWorkset.hess_vec_prod_f_xx.is_null();
+  bool f_xp_is_active = !workset.hessianWorkset.hess_vec_prod_f_xp.is_null();
+  bool f_px_is_active = !workset.hessianWorkset.hess_vec_prod_f_px.is_null();
+
+  const bool is_active = g_xx_is_active || g_xp_is_active || f_xx_is_active || f_xp_is_active;
```
Comment (mperego) [inline src/LandIce/evaluators/LandIce_Gather2DField_Def.hpp:370]: Can you add a comment explaining what `is_active` and `is_direction_active` are?

File: src/LandIce/evaluators/LandIce_GatherVerticallyContractedSolution_Def.hpp [hunk 2]
```diff
@@ -416,8 +416,14 @@ evaluateFields(typename Traits::EvalData workset)
   bool g_xx_is_active = !workset.hessianWorkset.hess_vec_prod_g_xx.is_null();
   bool g_xp_is_active = !workset.hessianWorkset.hess_vec_prod_g_xp.is_null();
   bool g_px_is_active = !workset.hessianWorkset.hess_vec_prod_g_px.is_null();
+  bool f_xx_is_active = !workset.hessianWorkset.hess_vec_prod_f_xx.is_null();
+  bool f_xp_is_active = !workset.hessianWorkset.hess_vec_prod_f_xp.is_null();
+  bool f_px_is_active = !workset.hessianWorkset.hess_vec_prod_f_px.is_null();
 
-  if(g_xx_is_active||g_px_is_active) {
+  const bool is_active = g_xx_is_active || g_xp_is_active || f_xx_is_active || f_xp_is_active;
```
Comment (mperego) [inline src/LandIce/evaluators/LandIce_GatherVerticallyContractedSolution_Def.hpp:423]: same here. Can you add a comment explaining what these variables are?

File: src/evaluators/bc/PHAL_Dirichlet.hpp [hunk 3]
```diff
@@ -120,6 +120,7 @@ class Dirichlet<PHAL::AlbanyTraits::HessianVec,Traits>
    : public DirichletBase<PHAL::AlbanyTraits::HessianVec, Traits> {
 public:
   Dirichlet(Teuchos::ParameterList& p);
+  void preEvaluate(typename Traits::EvalData d);
```
Comment (mperego) [inline src/evaluators/bc/PHAL_Dirichlet.hpp:123]: please add a comment briefly explaining why we need pre-evaluate.

File: src/evaluators/bc/PHAL_DirichletField.hpp [hunk 4]
```diff
@@ -102,6 +102,7 @@ class DirichletField<PHAL::AlbanyTraits::HessianVec, Traits>
   public:
     DirichletField(Teuchos::ParameterList& p);
     typedef typename PHAL::AlbanyTraits::HessianVec::ScalarT ScalarT;
+    void preEvaluate(typename Traits::EvalData d);
```
Comment (mperego) [inline src/evaluators/bc/PHAL_DirichletField.hpp:105]: please add a comment briefly explaining why we need preEvaluate.

File: src/evaluators/bc/PHAL_SDirichlet.hpp [hunk 5]
```diff
@@ -124,6 +124,9 @@ class SDirichlet<PHAL::AlbanyTraits::HessianVec, Traits>
 
   SDirichlet(Teuchos::ParameterList& p);
 
+  void
+  preEvaluate(typename Traits::EvalData d);
```
Comment (mperego) [inline src/evaluators/bc/PHAL_SDirichlet.hpp:128]: please add a comment briefly explaining why we need preEvaluate.

File: src/evaluators/bc/PHAL_SDirichletField.hpp [hunk 6]
```diff
@@ -127,6 +127,7 @@ class SDirichletField<PHAL::AlbanyTraits::HessianVec, Traits>
 
     SDirichletField(Teuchos::ParameterList& p);
 
+    void preEvaluate(typename Traits::EvalData d);
```
Comment (mperego) [inline src/evaluators/bc/PHAL_SDirichletField.hpp:130]: please add a comment briefly explaining why we need prEvaluate.

File: src/evaluators/gather/PHAL_GatherScalarNodalParameter_Def.hpp [hunk 7]
```diff
@@ -361,10 +361,13 @@ evaluateFields(typename Traits::EvalData workset)
...
   bool g_px_is_active = !workset.hessianWorkset.hess_vec_prod_g_px.is_null();
   bool g_pp_is_active = !workset.hessianWorkset.hess_vec_prod_g_pp.is_null();
+  bool f_xp_is_active = !workset.hessianWorkset.hess_vec_prod_f_xp.is_null();
+  bool f_px_is_active = !workset.hessianWorkset.hess_vec_prod_f_px.is_null();
+  bool f_pp_is_active = !workset.hessianWorkset.hess_vec_prod_f_pp.is_null();
 
   // Are we differentiating w.r.t. this parameter?
-  const bool is_active = (workset.dist_param_deriv_name == this->param_name) && (g_px_is_active||g_pp_is_active);
-  const bool is_direction_active = (workset.hessianWorkset.dist_param_deriv_direction_name == this->param_name) && (g_xp_is_active || g_pp_is_active);
+  const bool is_active = (workset.dist_param_deriv_name == this->param_name) && (g_px_is_active||g_pp_is_active||f_px_is_active||f_pp_is_active);
```
Comment (mperego) [inline src/evaluators/gather/PHAL_GatherScalarNodalParameter_Def.hpp:369]: please add a comment briefly explaining what `is_active` and `is_direction_active` are

File: src/evaluators/gather/PHAL_GatherScalarNodalParameter_Def.hpp [hunk 8]
```diff
@@ -439,10 +442,13 @@ evaluateFields(typename Traits::EvalData workset)
...
```
Comment (mperego) [inline src/evaluators/gather/PHAL_GatherScalarNodalParameter_Def.hpp:450]: same here, please add comment

File: src/evaluators/scatter/PHAL_ScatterResidual.hpp [hunk 9]
```diff
@@ -290,6 +290,7 @@ class ScatterResidualWithExtrudedParams<PHAL::AlbanyTraits::HessianVec,Traits>
                       PHX::FieldManager<Traits>& vm) {
     ScatterResidual<PHAL::AlbanyTraits::HessianVec, Traits>::postRegistrationSetup(d,vm);
   }
+  void evaluate2DFieldsDerivativesDueToExtrudedParams(typename Traits::EvalData d);
```
Comment (mperego) [inline src/evaluators/scatter/PHAL_ScatterResidual.hpp:293]: Is this used? (we can chat about that later)

**Review by mperego — COMMENTED** [review]
Looks good to me. Thanks! 
I left a few comments. In addition to those, can you add a brief comment where you declare the  evaluators in the Gather and Scatter for the HessianVec type, briefly explaining what you are doing. Something similar to what you did in the text of this PR.

**Comment by mperego** [comment]
@ikalash I mainly asked @kliegeois to add some comments. In the mean time I think you can test it, if you like to.

**Comment by ikalash** [comment]
I checked this in a couple of my nightlies, and I think it is good to merge in.  Should issues crop up in some of the other nightlies, we can fix them later.

**Comment by mperego** [comment]
> I checked this in a couple of my nightlies, and I think it is good to merge in. Should issues crop up in some of the other nightlies, we can fix them later.

OK, thanks for checking!

**Comment by kliegeois** [comment]
I am currently working on the comments, I will push them today.

**Comment by mperego** [comment]
@kliegeois one more thing. If you don't mind, can you please rebase and clean up a bit the commits?

**Comment by kliegeois** [comment]
@mperego, I will do that, thanks!

**Comment by mperego** [comment]
> @mperego, I will do that, thanks!

Thanks!

**Comment by kliegeois** [comment]
@mperego , please do not merge this PR today, I observed an issue with one of my MPI test.
I am on it.

**Comment by kliegeois** [comment]
@mperego , in the meantime, I rebased the branch to squash the commits and added the comments as requested.
I verified that those comments can be compiled with Doxygen, can you let me know if they are sufficiently clear to you?
Thanks a lot!

**Review by mperego — APPROVED** [review]
Comments are great. Thanks!

**Comment by kliegeois** [comment]
@mperego , thanks for your review. I found the bug, I am correcting it. I test it and I let you know when the PR can be merged!

**Comment by kliegeois** [comment]
@mperego , my last commit [3b649dc1c313519671f086778295ce422ed6acab](https://github.com/SNLComputation/Albany/pull/637/commits/3b649dc1c313519671f086778295ce422ed6acab) solved the issue.
The PR can be merged!
Thanks.

**Comment by mperego** [comment]
> @mperego , my last commit [3b649dc1c313519671f086778295ce422ed6acab](https://github.com/SNLComputation/Albany/pull/637/commits/3b649dc1c313519671f086778295ce422ed6acab) solved the issue.
> The PR can be merged!
> Thanks.

Merged! Thanks.

**Pull request merged**